File: scripts/trading_launcher_mcp.js
#!/usr/bin/env node
// MCP server that launches and validates the Jarvis trading web UI.

const { spawn } = require("child_process");
const fs = require("fs");
const path = require("path");
const readline = require("readline");

const ROOT = process.env.JARVIS_ROOT || path.resolve(__dirname, "..");
const PYTHON = process.env.JARVIS_TRADING_PYTHON || "python";
const LOG_DIR = path.join(ROOT, "logs");
const LOG = path.join(LOG_DIR, "trading_web.log");

fs.mkdirSync(LOG_DIR, { recursive: true });

let flask = null;

function readLogTail(maxLines = 25) {
  try {
    const text = fs.readFileSync(LOG, "utf8");
    const lines = text.split(/\r?\n/).filter(Boolean);
    return lines.slice(-maxLines).join("\n") || "(log file is empty)";
  } catch (error) {
    if (error && error.code === "ENOENT") return "(log file not found)";
    return `(unable to read log: ${error && error.message ? error.message : String(error)})`;
  }
}

function send(obj) {
  process.stdout.write(JSON.stringify(obj) + "\n");
}

function sendError(id, code, message, data) {
  const payload = { jsonrpc: "2.0", id, error: { code, message } };
  if (data) payload.error.data = data;
  send(payload);
}

function isAlive() {
  return !!(flask && flask.exitCode === null);
}

async function ensureFlask() {
  if (isAlive()) {
    return {
      ok: true,
      details: { status: "running", url: "http://127.0.0.1:5001", pid: flask.pid, logPath: LOG },
    };
  }

  const out = fs.openSync(LOG, "w");
  let spawnError = null;
  flask = spawn(PYTHON, [path.join(ROOT, "web", "trading_web.py")], {
    cwd: ROOT,
    stdio: ["ignore", out, out],
  });
  flask.once("error", (err) => {
    spawnError = err;
  });

  await new Promise((resolve) => setTimeout(resolve, 600));

  if (spawnError || !isAlive()) {
    return {
      ok: false,
      details: {
        status: "failed",
        exitCode: flask ? flask.exitCode : null,
        error: spawnError ? String(spawnError.message || spawnError) : null,
        logPath: LOG,
        logTail: readLogTail(),
      },
    };
  }

  return {
    ok: true,
    details: { status: "running", url: "http://127.0.0.1:5001", pid: flask.pid, logPath: LOG },
  };
}

function statusPayload() {
  if (isAlive()) {
    return { status: "running", url: "http://127.0.0.1:5001", pid: flask.pid, logPath: LOG };
  }
  if (!flask) {
    return { status: "stopped", logPath: LOG, logTail: readLogTail() };
  }
  return { status: "failed", exitCode: flask.exitCode, logPath: LOG, logTail: readLogTail() };
}

const rl = readline.createInterface({ input: process.stdin, crlfDelay: Infinity });

rl.on("line", async (line) => {
  if (!line.trim()) return;
  let msg;
  try {
    msg = JSON.parse(line);
  } catch {
    return;
  }

  const { method, id } = msg;

  if (method === "initialize") {
    send({
      jsonrpc: "2.0",
      id,
      result: {
        protocolVersion: "2024-11-05",
        capabilities: { tools: {} },
        serverInfo: { name: "trading-launcher", version: "1.1.0" },
      },
    });
    return;
  }

  if (method === "notifications/initialized") {
    return;
  }

  if (method === "tools/list") {
    send({
      jsonrpc: "2.0",
      id,
      result: {
        tools: [
          {
            name: "trading_status",
            description: "Returns Jarvis Trading UI status and URL when healthy.",
            inputSchema: { type: "object", properties: {} },
          },
        ],
      },
    });
    return;
  }

  if (method === "tools/call") {
    const result = await ensureFlask();
    if (!result.ok) {
      sendError(id, -32000, "Trading UI failed to start", result.details);
      return;
    }

    send({
      jsonrpc: "2.0",
      id,
      result: {
        content: [
          { type: "text", text: "Trading UI is running at http://127.0.0.1:5001" },
          { type: "text", text: JSON.stringify(result.details) },
        ],
      },
    });
    return;
  }

  if (method === "trading/status") {
    send({ jsonrpc: "2.0", id, result: statusPayload() });
    return;
  }

  if (id !== undefined) {
    sendError(id, -32601, "Method not found");
  }
});

File: scripts/trading_launcher_mcp.py
"""MCP server that launches and validates the Jarvis trading web UI."""

from __future__ import annotations

import json
import os
import subprocess
import sys
import time
from typing import Any

DEFAULT_ROOT = r"C:\Users\lucid\Desktop\Jarvis"
ROOT = os.environ.get("JARVIS_ROOT", DEFAULT_ROOT)
PYTHON = os.environ.get("JARVIS_TRADING_PYTHON", "python")
LOG_PATH = os.path.join(ROOT, "logs", "trading_web.log")

os.makedirs(os.path.join(ROOT, "logs"), exist_ok=True)

_proc: subprocess.Popen[str] | None = None
_log_handle = None


def _read_log_tail(max_lines: int = 25) -> str:
    try:
        with open(LOG_PATH, "r", encoding="utf-8", errors="replace") as fh:
            lines = fh.readlines()
    except FileNotFoundError:
        return "(log file not found)"
    except Exception as exc:  # pragma: no cover - defensive path
        return f"(unable to read log: {exc})"

    return "".join(lines[-max_lines:]).strip() or "(log file is empty)"


def _send(obj: dict[str, Any]) -> None:
    line = json.dumps(obj, ensure_ascii=False) + "\n"
    os.write(sys.stdout.fileno(), line.encode("utf-8"))


def _send_error(mid: Any, code: int, message: str, data: dict[str, Any] | None = None) -> None:
    payload: dict[str, Any] = {"jsonrpc": "2.0", "id": mid, "error": {"code": code, "message": message}}
    if data is not None:
        payload["error"]["data"] = data
    _send(payload)


def _is_alive() -> bool:
    return _proc is not None and _proc.poll() is None


def ensure_flask() -> tuple[bool, dict[str, Any] | None]:
    """Start Flask when needed and verify it did not crash immediately."""
    global _proc, _log_handle
    if _is_alive():
        return True, None

    if _log_handle is not None:
        try:
            _log_handle.close()
        except Exception:
            pass
        _log_handle = None

    _log_handle = open(LOG_PATH, "w", encoding="utf-8", errors="replace")
    try:
        _proc = subprocess.Popen(
            [PYTHON, os.path.join(ROOT, "web", "trading_web.py")],
            stdout=_log_handle,
            stderr=_log_handle,
            cwd=ROOT,
        )
    except OSError as exc:
        return False, {
            "status": "failed",
            "exitCode": None,
            "logPath": LOG_PATH,
            "error": str(exc),
            "logTail": _read_log_tail(),
        }

    # Allow immediate import/runtime failures to surface before reporting "running".
    time.sleep(0.6)
    if _proc.poll() is not None:
        return False, {
            "status": "failed",
            "exitCode": _proc.returncode,
            "logPath": LOG_PATH,
            "logTail": _read_log_tail(),
        }

    return True, {"status": "running", "url": "http://127.0.0.1:5001", "pid": _proc.pid}


def _status_payload() -> dict[str, Any]:
    if _is_alive():
        assert _proc is not None
        return {
            "status": "running",
            "url": "http://127.0.0.1:5001",
            "pid": _proc.pid,
            "logPath": LOG_PATH,
        }

    if _proc is None:
        return {
            "status": "stopped",
            "logPath": LOG_PATH,
            "logTail": _read_log_tail(),
        }

    return {
        "status": "failed",
        "exitCode": _proc.returncode,
        "logPath": LOG_PATH,
        "logTail": _read_log_tail(),
    }


for raw in sys.stdin.buffer:
    try:
        msg = json.loads(raw)
        method, mid = msg.get("method", ""), msg.get("id")

        if method == "initialize":
            _send(
                {
                    "jsonrpc": "2.0",
                    "id": mid,
                    "result": {
                        "protocolVersion": "2024-11-05",
                        "capabilities": {"tools": {}},
                        "serverInfo": {"name": "trading-launcher", "version": "1.1.0"},
                    },
                }
            )
            continue

        if method == "tools/list":
            _send(
                {
                    "jsonrpc": "2.0",
                    "id": mid,
                    "result": {
                        "tools": [
                            {
                                "name": "trading_status",
                                "description": "Get Jarvis Trading UI status and URL if healthy.",
                                "inputSchema": {"type": "object", "properties": {}},
                            }
                        ]
                    },
                }
            )
            continue

        if method == "tools/call":
            ok, details = ensure_flask()
            if not ok:
                _send_error(
                    mid,
                    -32000,
                    "Trading UI failed to start",
                    details,
                )
                continue

            _send(
                {
                    "jsonrpc": "2.0",
                    "id": mid,
                    "result": {
                        "content": [
                            {
                                "type": "text",
                                "text": "Trading UI is running at http://127.0.0.1:5001",
                            },
                            {"type": "text", "text": json.dumps(details, ensure_ascii=False)},
                        ]
                    },
                }
            )
            continue

        if method == "trading/status":
            _send({"jsonrpc": "2.0", "id": mid, "result": _status_payload()})
            continue

        if mid is not None:
            _send_error(mid, -32601, "Method not found")
    except Exception:
        # Keep server loop alive for malformed client payloads.
        continue
